File: Basics/3.polymorphism.cpp
#include <bits/stdc++.h>
using namespace std;

/*
    Polymorphism: ability of a function to perform different tasks based on the object it is acting upon.
    - Compile Time Polymorphism: function overloading, operator overloading.
    - Run Time Polymorphism: function overriding using virtual functions.
*/

class Shape {
public:
    // Virtual function for runtime polymorphism
    virtual void area() {
        cout << "Calculating area of shape." << endl;
    }
};

class Circle : public Shape {
private:
    double radius;

public:
    // Constructor overloading (compile time polymorphism)
    Circle() {}

    Circle(double r) {
        radius = r;
    }

    void area() override {
        cout << "Area of Circle: " << 3.1416 * radius * radius << endl;
    }
};

class Rectangle : public Shape {
private:
    double length, width;

public:
    Rectangle(double l, double w) {
        length = l;
        width = w;
    }

    void area() override {
        cout << "Area of Rectangle: " << length * width << endl;
    }
};

int main() {
    // Compile time polymorphism through constructor overloading
    Circle c1(5.0);
    c1.area();

    Rectangle r1(4.0, 6.0);
    r1.area();

    // Run time polymorphism through base class pointer
    Shape* shape;

    shape = &c1;
    shape->area();

    shape = &r1;
    shape->area();

    return 0;
}

File: 1.Basics/3.polymorphism.java
import java.util.*;

/*
    Polymorphism: ability of a function to perform different tasks based on the object it is acting upon.
    - Compile Time Polymorphism: method overloading, operator overloading.
    - Run Time Polymorphism: method overriding using virtual functions.
*/

abstract class Shape {
    // Abstract method for runtime polymorphism
    abstract void area();
}

class Circle extends Shape {
    private double radius;

    // Constructor overloading (compile time polymorphism)
    Circle() {}

    Circle(double r) {
        radius = r;
    }

    @Override
    void area() {
        System.out.println("Area of Circle: " + 3.1416 * radius * radius);
    }
}

class Rectangle extends Shape {
    private double length, width;

    Rectangle(double l, double w) {
        length = l;
        width = w;
    }

    @Override
    void area() {
        System.out.println("Area of Rectangle: " + length * width);
    }
}

public class Main {
    public static void main(String[] args) {
        // Compile time polymorphism through constructor overloading
        Circle c1 = new Circle(5.0);
        c1.area();

        Rectangle r1 = new Rectangle(4.0, 6.0);
        r1.area();

        // Run time polymorphism through base class reference
        Shape shape;

        shape = c1;
        shape.area();

        shape = r1;
        shape.area();
    }
}
